### ISSUE TITLE
Bug: Savefile Corruption

### ISSUE THREAD

**Issue by ThatGitMITZE** [NickMcConnell/NarSil#824]
Okay, I'm not entirely positive about this one, it doesn't look like there's been recent pushed changes, so it shouldn't be a savefile backwards compatibility thing, but I'm not so sure it's not specifically a webserver interaction issue.

---

Webserver nightly, I've been having sporadic issues where, upon dying and closing out of the game, something seems to happen that corrupts the savefile—I don't know what, the game seems to end fine, but when I go to start a new game and the webserver tries to automatically load my user savefile, it gives me:

Error reading item
Savefile corrupted - Couldn't load block gear

It doesn't happen every run, but it's happened to me a few times now, and I have no leads on if I'm doing something to cause it—I'm not letting the webserver time me out, I'm ending the game as it should expect to end normally.

**Comment by NickMcConnell** [comment]
If you want to download and send me the savefile I can have a look

**Comment by ThatGitMITZE** [comment]
[MITZE.zip](https://github.com/user-attachments/files/18243439/MITZE.zip)

Here's (what I think is) the savefile; not entirely positive I have the right thing because of how the server operates. And yes, that's an actual .zip, not the savefile renamed to have a .zip extension.

FWIW BTW I'm less concerned about savefile recovery (I'm not making nearly enough progress to care about that) and more about whatever's happening.

**Comment by NickMcConnell** [comment]
That was indeed the savefile, and I can confirm the issue.  On first analysis it looks like knowledge is sometimes not being transferred fully to the known object - the problem was a known object with no tval or sval, but pd = ps = 1 (possibly a cloak of protection).

I don't suppose you have noticed, say, always having a Protection ego when this happens, or any other carried/worn item ?

**Comment by ThatGitMITZE** [comment]
I am often making Cloaks of Protection, trying to get smith starts off the ground: 6-7 Smithing, Armorsmith, Enchantment, Gloves of the Forge, Crown of Grace, leftover use that I toss into a Cloak of Protection for a little bit more EV and PROT.

Not that it's been helping.

**Comment by NickMcConnell** [comment]
OK, so it might be smithing still not quite right then.

**Comment by backwardsEric** [comment]
It's the known version of a pair of boots that's failing to load.  According to the corresponding actual object in the gear list, those boots are fairly ordinary (artifact and ego fields are NULL; pd and ps are one, however) and were dropped by an orc at 100 feet.

**Comment by NickMcConnell** [comment]
Thanks for chasing that down, I have barely looked

**Comment by NickMcConnell** [comment]
Just happened to me on my .live game too.

**Comment by NickMcConnell** [comment]
In my case it seems to be a Battle Axe, dropped by an orc scout at 150'.  The first thing that occurs to me is that object_set_base_known() hasn't been called on the object, but I can't see why.

**Comment by backwardsEric** [comment]
Vanilla Angband has object_see() call object_set_base_known() when the known version either does not exist or the its kind differs from that of the actual object.  In NarSil, it is only setting the kind and number in those cases.  In the case where a drop from a monster ends up at the character's feet that sets up a path where the known version does not have the tval and sval set.

drop_near() causes square_note_spot() to be called which calls object_see() and then object_touch() for any objects on the character's grid.  object_see() sets the kind and number for the known version.  object_touch() sets OBJ_NOTICE_ASSESSED on the known version and then calls player_know_object().  player_know_object() only calls object_set_base_known() when OBJ_NOTICE_ASSESSED is not set so that known object would not pick up the base information on any future calls to player_know_object().

There's paths for other objects on the floor that have been seen but not touched to pick up the base information if update_player_object_knowledge() happens to called.  At a minimum, that happens whenever the player's bonuses are updated.

**Comment by NickMcConnell** [comment]
Yes, I had a look at that and thought it looked wrong, then compared with Vanilla, and then forgot about it all...

I think the whole setting of OBJ_NOTICE_ASSESSED needs a redo.

**Comment by NickMcConnell** [comment]
Note that 79331081 may help this too.

**Comment by backwardsEric** [comment]
Without the object_see() change, picking up a drop that had rolled under the character's feet, exiting the game, and then reloading did trigger the error on reloading and the offending item was the picked up drop.  After the change, doing the same reloads fine so at least one source of the corruption has been plugged.

**Comment by NickMcConnell** [comment]
OK, I'm going to claim that any future corruption is a new issue and close this one.
